slides/intro-extra-recipes.qmd

```diff
@@ -0,0 +1,378 @@
+\ \ step_zv([all_predictors()]{style="color: #CA225E;"}) %>%  
+\ \ step_log([distance]{style="color: #CA225E;"}) %>% 
+\ \ step_normalize([all_numeric_predictors()]{style="color: #CA225E;"})
+:::
+<br>
+
+Use {tidyselect} and recipes-specific selectors to denote affected variables.
+
+## Using a recipe
+
+:::{style="font-family: 'Source Code Pro', monospace; font-size: 0.8em;"}
+rec_spec <- recipe(tip ~ ., data = taxi_train) %>%  
+\ \ step_unknown(all_nominal_predictors()) %>%  
+\ \ step_dummy(all_nominal_predictors()) %>%  
+\ \ step_zv(all_predictors()) %>%  
+\ \ step_log(distance) %>% 
+\ \ step_normalize(all_numeric_predictors())
+:::
+
+Save the recipe we like
+
+<br>
+
+## Using a recipe with workflows
+
+recipes are can be used with {workflows} to "combine" it with a model
+
+<br>
+
+:::{style="font-family: 'Source Code Pro', monospace; font-size: 0.8em;"}
+wf_spec <- workflow() %>%  
+\ \ [add_recipe(rec_spec)]{style="color: #CA225E;"} %>%  
+\ \ add_model(linear_reg())
+:::
+
+## Recipes are estimated
+
+<br>
+
+Every preprocessing step in a recipe that involved calculations uses the *training* set. For example:
+
+- Levels of a factor
+- Determination of zero-variance
+- Normalization
+- Feature extraction
+
+Once a a recipe is added to a workflow, this occurs when `fit()` is called.
+
+
+## Debugging a recipe
+
+- Typically, you will want to use a workflow to estimate and apply a recipe.
+
+. . .
+
+- If you have an error and need to debug your recipe, the original recipe object (e.g. `rec_spec`) can be estimated manually with a function called `prep()`. It is analogous to `fit()`. See [TMwR section 16.4](https://www.tmwr.org/dimensionality.html#recipe-functions)
+
+. . .
+
+- Another function (`bake()`) is analogous to `predict()`, and gives you the processed data back.
+
+. . .
+
+- The `tidy()` function can be used to get specific results from the recipe.
+
+## Your turn {transition="slide-in"}
+
+![](images/parsnip-flagger.jpg){.absolute top="0" right="0" width="150" height="150"}
+
+<br>
+
+*Take the recipe and `prep()` then `bake()` it to see what the resulting data set looks like.*
+
+*Try removing steps to see how the result changes.*
+
+<br>
+
+```{r}
+#| label: recipes-prep-exercise
+#| echo: false
+countdown(minutes = 5, id = "recipes-prep")
+```
+
+## Printing a recipe
+
+```{r}
+#| label: rec_spec
+rec_spec <- recipe(tip ~ ., data = taxi_train) %>%  
+  step_unknown(all_nominal_predictors()) %>%  
+  step_dummy(all_nominal_predictors()) %>%  
+  step_zv(all_predictors()) %>%  
+  step_log(distance) %>% 
+  step_normalize(all_numeric_predictors())
+```
+
+```{r}
+#| label: rec_spec-printing
+#| message: true
+rec_spec
+```
+
+## Prepping a recipe
+
+```{r}
+#| label: rec_spec-prep-printing
+#| message: true
+prep(rec_spec)
+```
+
+## Baking a recipe
+
+```{r}
+#| label: rec_spec-prep-bake-printing
+prep(rec_spec) %>%
+  bake(new_data = taxi_train)
```

Comment:
`distance` is all `NaN`--having trouble putting my finger on this, but seems to be due to `step_normalize()`:

``` r
library(tidymodels)
library(modeldatatoo)

recipe_base <-
  recipe(tip ~ ., data = data_taxi()) %>%  
  step_unknown(all_nominal_predictors()) %>%  
  step_dummy(all_nominal_predictors()) %>%  
  step_zv(all_predictors()) %>%  
  step_log(distance) 

recipe_base %>% 
  prep() %>% 
  bake(NULL)
#> # A tibble: 10,000 × 19
#>    distance  hour tip   company_City.Service company_Flash.Cab company_Sun.Taxi
#>       <dbl> <int> <fct>                <dbl>             <dbl>            <dbl>
#>  1   2.84      16 yes                      0                 0                0
#>  2  -0.128      8 yes                      1                 0                0
#>  3   2.90      18 yes                      0                 0                0
#>  4   3.03       8 yes                      0                 0                0
#>  5   2.50      21 yes                      0                 0                0
#>  6  -0.0619    23 yes                      0                 0                1
#>  7   2.86      12 yes                      0                 1                0
#>  8   2.87       6 yes                      0                 0                0
#>  9   0.615     12 yes                      0                 0                0
#> 10   0.385     14 yes                      1                 0                0
#> # ℹ 9,990 more rows
#> # ℹ 13 more variables: company_Taxi.Affiliation.Services <dbl>,
#> #   company_Taxicab.Insurance.Agency.Llc <dbl>, company_other <dbl>,
#> #   local_no <dbl>, dow_Mon <dbl>, dow_Tue <dbl>, dow_Wed <dbl>, dow_Thu <dbl>,
#> #   dow_Fri <dbl>, dow_Sat <dbl>, month_Feb <dbl>, month_Mar <dbl>,
#> #   month_Apr <dbl>
  
recipe_base %>%
  step_normalize(all_numeric_predictors()) %>%
  prep() %>%
  bake(NULL)
#> # A tibble: 10,000 × 19
#>    distance    hour tip   company_City.Service company_Flash.Cab
#>       <dbl>   <dbl> <fct>                <dbl>             <dbl>
#>  1      NaN  0.418  yes                 -0.367            -0.335
#>  2      NaN -1.42   yes                  2.72             -0.335
#>  3      NaN  0.877  yes                 -0.367            -0.335
#>  4      NaN -1.42   yes                 -0.367            -0.335
#>  5      NaN  1.56   yes                 -0.367            -0.335
#>  6      NaN  2.02   yes                 -0.367            -0.335
#>  7      NaN -0.499  yes                 -0.367             2.98 
#>  8      NaN -1.88   yes                 -0.367            -0.335
#>  9      NaN -0.499  yes                 -0.367            -0.335
#> 10      NaN -0.0407 yes                  2.72             -0.335
#> # ℹ 9,990 more rows
#> # ℹ 14 more variables: company_Sun.Taxi <dbl>,
#> #   company_Taxi.Affiliation.Services <dbl>,
#> #   company_Taxicab.Insurance.Agency.Llc <dbl>, company_other <dbl>,
#> #   local_no <dbl>, dow_Mon <dbl>, dow_Tue <dbl>, dow_Wed <dbl>, dow_Thu <dbl>,
#> #   dow_Fri <dbl>, dow_Sat <dbl>, month_Feb <dbl>, month_Mar <dbl>,
#> #   month_Apr <dbl>
```

<sup>Created on 2023-07-26 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

Reply:
whoops, there are some 0 distances, i fixed it by setting `offset = 0.5` in `step_log()`

Reply:
[`2a3dc45` (#233)](https://github.com/tidymodels/workshops/pull/233/commits/2a3dc457f00211a1ec75383d90785bdd3a692d1b)